src/backend/common/kernel_cache.cpp

```diff
@@ -68,15 +68,15 @@ Kernel getKernel(const string& kernelName,
         tInstance += '>';
     }
 
-    const bool notJIT = !sourceIsJIT;
-    size_t moduleKey;
-    if (notJIT) {
+    size_t moduleKey = 0;
+    if (sourceIsJIT) {
+        moduleKey = deterministicHash(tInstance);
+    } else {
         moduleKey = (sources.size() == 1 && sources[0].hash)
                         ? sources[0].hash
                         : deterministicHash(sources);
         moduleKey = deterministicHash(options, moduleKey);
-    } else {
-        moduleKey = deterministicHash(tInstance);
+        moduleKey = deterministicHash(tInstance, moduleKey);
```

Comment:
I have the impression that this is only for CUDA.  Sorry, that I did not realize this before, and as a result broke the code.
For OpenCL, it has a negative effect on speed since now programs with multiple kernels are compiled multiple times (1 for each kernel) and time is avoidable time is spend in the hash.

I suggest that following changes:
- put the tInstance generation (loop) inside a #define(AF_CUDA) and only keep the initialization with kernelName.
- put the determiniticHash(tInstance,moduleKey) also after a #define(AF_CUDA)
- The call at the end, to getKernel can so become common for both.  no #define needed anymore.

During my tests, I could measure a performance difference for OpenCL.  By doing so, performance will even improve further, since this is on the critical path.

Reply:
1. Multiple compilations are not possible, let me explain why. For OpenCL backend, the template arguments are sort of dummy since they are also being supplied as part of compilation options because of the way OpenCL kernels work - they can be only written in C. Compilation options are different for identical named kernels depending on other opencl-compile time constants which can be types, constants depending on the need.

2. Earlier the bug wasn't just CUDA related. For example, `tInstance` is `approx<float>` in CUDA and just `approx` in OpenCL, a kernel file with identical content but with different kernel/function name would incorrectly fetch the wrong binary if we don't hash the name of the kernel. Although, this situation doesn't happen now, I don't think we should add this unnecessary limitation to the caching API.

3. It is for the reason mentioned in (1), we need to call `getKernel` in the end with different parameters. If we always pass `tInstance` in the end, we would end up moving the condition to the beginning to make sure OpenCL backend doesn't add `< ... >` to `tInstance` - I don't have any objection for this though since we end up doing less work adding those extra template parameters to `tInstance` which is not needed for OpenCL.

How much time difference are you noticing ? 

In another note, I think we should get rid of template args in OpenCL backend as in pass `{}` instead of any values which save some time in creating that variable all together.

Reply:
I have pushed the change from point (3) in earlier comment, can you please test the branch now for timing.

Reply:
@9prady9 

I have finally some testing results.

Scenario: I have 4 scenario's.  disk caching ON/OFF; deterministicHash(tInstance) added for OpenCL ON/OFF (code diff attached).

For ORB, I measure the first run (incl. compiling, storing & execute) and a loop (only searching & execute)

Disk caching | hash | 1st run | %time | +speed | Loop (ORB) | %time | +speed | # kernels safed
OFF | ON | 1.8465s | 100% | 1.0x | 0.0120s | 100% | 1.0x| 9 programs
OFF | OFF | 0.1885s | __10%__ | 10.0x | 0.0120s | 100% | 1.0x | __4 programs__
ON | ON | 0.2036s | 100% | 1.0x | 0.0119s | 100% | 1.0x | 9 programs
ON | OFF | 0.1887s | __93%__ | 1.1x |0.0119s | 100% | 1.0x | __4 programs__

These differences will be observed for:  fast, flood_fill, harris, homography, regions and orb.
Nice gains for the above functions during initialization, although it is only once for a whole program run.
For all the others the difference is within the error margin.

=======
An update on the neural_network test:
The final speed is very similar as before.  a 1% difference however between the hash(kernName) ON/OFF.  This is a general situation, without multiply kernels in 1 program source file.


=========
I get an error while attaching files, so I include them here in clear. 

Epochs:250; | Precision:f32 |   |   |   |   |   |   |   |   |   |   |   |   |   |   | speed vs master
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Batch | size: | 32 | \| | Training | time: | 37.2885 | s | \| | Accuracy | training: | 98.8924 | \| | Accuracy | test: | 92.4829 |   | 1.20
Batch | size: | 40 | \| | Training | time: | 30.2542 | s | \| | Accuracy | training: | 86.5763 | \| | Accuracy | test: | 82.7892 |   | 1.16
Batch | size: | 48 | \| | Training | time: | 24.4361 | s | \| | Accuracy | training: | 98.8263 | \| | Accuracy | test: | 93.3435 |   | 1.19
Batch | size: | 52 | \| | Training | time: | 23.7057 | s | \| | Accuracy | training: | 98.9254 | \| | Accuracy | test: | 93.065 |   | 1.13
Batch | size: | 64 | \| | Training | time: | 19.0509 | s | \| | Accuracy | training: | 99.0908 | \| | Accuracy | test: | 93.6219 |   | 1.17
Batch | size: | 96 | \| | Training | time: | 15.4451 | s | \| | Accuracy | training: | 98.9089 | \| | Accuracy | test: | 93.3688 |   | 0.99
Batch | size: | 128 | \| | Training | time: | 13.1447 | s | \| | Accuracy | training: | 98.3138 | \| | Accuracy | test: | 92.9132 |   | 0.99
Batch | size: | 256 | \| | Training | time: | 10.4702 | s | \| | Accuracy | training: | 95.5034 | \| | Accuracy | test: | 92.407 |   | 1.01
Batch | size: | 512 | \| | Training | time: | 8.58663 | s | \| | Accuracy | training: | 33.4435 | \| | Accuracy | test: | 33.2068 |   | 1.01
Batch | size: | 1024 | \| | Training | time: | 7.73802 | s | \| | Accuracy | training: | 13.7709 | \| | Accuracy | test: | 13.389 |   | 1.01

=================
$ git diff src/backend/common/kernel_cache.cpp
diff --git a/src/backend/common/kernel_cache.cpp b/src/backend/common/kernel_cache.cpp
index 43ca6893..5031d6b7 100644
--- a/src/backend/common/kernel_cache.cpp
+++ b/src/backend/common/kernel_cache.cpp
@@ -81,7 +81,9 @@ Kernel getKernel(const string& kernelName,
                         ? sources[0].hash
                         : deterministicHash(sources);
         moduleKey = deterministicHash(options, moduleKey);
+#if defined(AF_CUDA)
         moduleKey = deterministicHash(tInstance, moduleKey);
+#endif
     }
     const int device  = detail::getActiveDeviceId();
     Module currModule = findModule(device, moduleKey);